trino-aws-proxy/src/main/java/io/trino/aws/proxy/server/rest/RequestFilter.java

```diff
@@ -0,0 +1,198 @@
+ */
+package io.trino.aws.proxy.server.rest;
+
+import com.google.common.base.Throwables;
+import com.google.common.collect.ImmutableMap;
+import com.google.inject.Inject;
+import io.airlift.log.Logger;
+import io.trino.aws.proxy.spi.rest.Request;
+import io.trino.aws.proxy.spi.signing.SigningController;
+import io.trino.aws.proxy.spi.signing.SigningMetadata;
+import io.trino.aws.proxy.spi.signing.SigningServiceType;
+import jakarta.ws.rs.WebApplicationException;
+import jakarta.ws.rs.container.ContainerRequestContext;
+import jakarta.ws.rs.container.ContainerRequestFilter;
+import jakarta.ws.rs.container.ContainerResponseContext;
+import jakarta.ws.rs.container.ContainerResponseFilter;
+import org.glassfish.jersey.server.ContainerRequest;
+import org.glassfish.jersey.server.ContainerResponse;
+import org.glassfish.jersey.server.model.Parameter;
+import org.glassfish.jersey.server.spi.internal.ValueParamProvider;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Function;
+
+import static jakarta.ws.rs.core.Response.Status.BAD_REQUEST;
+import static java.util.Objects.requireNonNull;
+import static org.glassfish.jersey.server.spi.internal.ValueParamProvider.Priority.HIGH;
+
+public class RequestFilter
+        implements ContainerRequestFilter, ContainerResponseFilter, ValueParamProvider
+{
+    private final Logger log = Logger.get(RequestFilter.class);
+    private final SigningController signingController;
+    private final Map<Class<?>, SigningServiceType> serviceTypesMap;
+    private final RequestLoggerController requestLoggerController;
+
+    private record InternalRequestContext(Request request, SigningMetadata signingMetadata, RequestLoggingSession requestLoggingSession)
+    {
+        private InternalRequestContext
+        {
+            requireNonNull(request, "request is null");
+            requireNonNull(signingMetadata, "signingMetadata is null");
+            requireNonNull(requestLoggingSession, "requestLoggingSession is null");
+        }
+    }
+
+    @Inject
+    RequestFilter(SigningController signingController, Map<Class<?>, SigningServiceType> serviceTypesMap, RequestLoggerController requestLoggerController)
+    {
+        this.signingController = requireNonNull(signingController, "signingController is null");
+        this.serviceTypesMap = ImmutableMap.copyOf(serviceTypesMap);
+        this.requestLoggerController = requireNonNull(requestLoggerController, "requestLoggerController is null");
+    }
+
+    @SuppressWarnings("ThrowableNotThrown")
+    @Override
+    public void filter(ContainerRequestContext requestContext)
+            throws IOException
+    {
+        if (requestContext.getRequest() instanceof ContainerRequest containerRequest) {
+            Class<?> declaringClass = containerRequest.getUriInfo().getMatchedResourceMethod().getInvocable().getDefinitionMethod().getDeclaringClass();
```

Comment:
the docs for `getMatchedResourceMethod` mentions `Returns:
The matched resource method that was invoked or null if no resource method was invoked.`.

From my understanding, filters are called after resources are matched (except if the `@PreMatch` annotation is used). Should we check that the returned method is null>